docs/proposals/24-dynamic-node-groups-for-shoot-cluster-autoscaler.md

```diff
@@ -0,0 +1,338 @@
+
+```go
+for _, pool := range w.worker.Spec.Pools {
+		zoneLen := int32(len(pool.Zones))
+deploymentName = fmt.Sprintf("%s-%s-z%d", w.worker.Namespace, pool.Name, zoneIndex+1)
+    for zoneIndex, zone := range pool.Zones {
+      //...
+      	machineDeployments = append(machineDeployments, worker.MachineDeployment{
+          //...
+  				Name:                 deploymentName,
+				  Minimum:              worker.DistributeOverZones(zoneIdx, pool.Minimum, zoneLen),
+				  Maximum:              worker.DistributeOverZones(zoneIdx, pool.Maximum, zoneLen),
+				  MaxSurge:             worker.DistributePositiveIntOrPercent(zoneIdx, pool.MaxSurge, zoneLen, pool.Maximum),
+				  MaxUnavailable:       worker.DistributePositiveIntOrPercent(zoneIdx, pool.MaxUnavailable, zoneLen, pool.Minimum),
+				  Labels:               addTopologyLabel(pool.Labels, zone),
+				  Annotations:          pool.Annotations,
+          //...
+			})
+    }
+}
+```
+
+##### MachineDeployment Min/Max Distribution 
+
+As can be seen from the snippet above, the machine deployment `Minimum` and `Maximum` delegates to the helper method `DistributeOverZones` which is coded as below:
+
+```go
+func DistributeOverZones(zoneIndex, size, zoneSize int32) int32 {
+	first := size / zoneSize
+	second := int32(0)
+	if zoneIndex < (size % zoneSize) {
+		second = 1
+	}
+	return first + second
+}
+```
+Even machine deployment `MaxSurge` and `MaxUnavailable` delegates to `DistributePositiveIntOrPercent` which in turn delegates to `DistributeOverZones`
+
+The above computation has issues as can be seen below for generated machine deployments.
+
+###### Pool-0:1:1:1:2 (min:max:maxsurge:maxunavail:numzones)
+```
+MachineDeployment_Z0(Minimum=0|Maximum:1|MaxSurge:1|MaxUnavailable:1)
+MachineDeployment_Z1(Minimum=0|Maximum:0|MaxSurge:0|MaxUnavailable:0)
+```
+Consider the case above where `numzones:2` > worker pool `max:1`, We can see that the machinedeployment for `Z1` is useless as its `Min`, `Max`, `MaxSurge` and `MaxUnavailable` are all zero making the corresponding CA `NodeGroup` also useless.
+
+
+###### Pool-3:5:1:1:2 (min:max:maxsurge:maxunavail:numzones)
+```
+MachineDeployment_Z0(Minimum=2|Maximum:3|MaxSurge:1|MaxUnavailable:1)
+MachineDeployment_Z1(Minimum=1|Maximum:2|MaxSurge:0|MaxUnavailable:0)
+```
+Consider the case above - `MaxSurge`  for `Z1` is computed as `0`
+
+The above cases are not un-common. Thre are `~65` Worker Pools in LIVE Landscape with `0 < WorkerPoolMax < WorkerPoolZones`
+
+
+
+
+### Deficiencies of Existing Solution
+
+1. The current static node group distribution skews distribution towards earlier specified zones over later specified zones.
+2. If zone length exceeds worker pool max, then later node groups get zero values effectively becoming useless.
+3. If the earlier node groups have quota issues, then later node groups have reduced capacity and cannot meet the specified availability. 
```

Comment:
Is it really about quota issues or just generally about instance capacity issues? I suspect it is rather the later, i.e. the infrastructure has problems providing new instances in one availability zone.

Reply:
changing from quota to capacity. it corresponds to CA `cloudprovider.OutOfResourcesErrorClass`